File: apps/macos/Kelpie/Handlers/ScreenshotHandler.swift
import AppKit

enum ScreenshotResolution: String {
    case native
    case viewport

    static func parse(_ raw: Any?) -> Self? {
        guard let raw else { return .native }
        guard let value = raw as? String else { return nil }
        return Self(rawValue: value)
    }
}

struct ScreenshotViewportMetrics {
    let viewportWidth: Int
    let viewportHeight: Int
    let devicePixelRatio: Double

    func metadata(
        imageWidth: Int,
        imageHeight: Int,
        format: String,
        resolution: ScreenshotResolution
    ) -> [String: Any] {
        let scaleX = viewportWidth > 0 ? Double(imageWidth) / Double(viewportWidth) : 1
        let scaleY = viewportHeight > 0 ? Double(imageHeight) / Double(viewportHeight) : 1
        return [
            "width": imageWidth,
            "height": imageHeight,
            "format": format,
            "resolution": resolution.rawValue,
            "coordinateSpace": "viewport-css-pixels",
            "viewportWidth": viewportWidth,
            "viewportHeight": viewportHeight,
            "devicePixelRatio": devicePixelRatio,
            "imageScaleX": scaleX,
            "imageScaleY": scaleY
        ]
    }
}

func bitmapRepresentation(of image: NSImage) -> NSBitmapImageRep? {
    if let data = image.tiffRepresentation,
       let bitmap = NSBitmapImageRep(data: data) {
        return bitmap
    }
    return nil
}

func scaledBitmapRepresentation(
    from image: NSImage,
    to resolution: ScreenshotResolution,
    using viewport: ScreenshotViewportMetrics
) -> NSBitmapImageRep? {
    guard let bitmap = bitmapRepresentation(of: image) else {
        return nil
    }
    guard resolution == .viewport else {
        return bitmap
    }
    let targetWidth = max(Int(round(Double(bitmap.pixelsWide) / max(viewport.devicePixelRatio, 1.0))), 1)
    let targetHeight = max(Int(round(Double(bitmap.pixelsHigh) / max(viewport.devicePixelRatio, 1.0))), 1)
    guard targetWidth != bitmap.pixelsWide || targetHeight != bitmap.pixelsHigh else {
        return bitmap
    }
    guard let scaled = NSBitmapImageRep(
        bitmapDataPlanes: nil,
        pixelsWide: targetWidth,
        pixelsHigh: targetHeight,
        bitsPerSample: bitmap.bitsPerSample,
        samplesPerPixel: bitmap.samplesPerPixel,
        hasAlpha: bitmap.hasAlpha,
        isPlanar: false,
        colorSpaceName: bitmap.colorSpaceName ?? .deviceRGB,
        bytesPerRow: 0,
        bitsPerPixel: 0
    ) else {
        return bitmap
    }
    scaled.size = NSSize(width: targetWidth, height: targetHeight)
    let source = NSImage(size: NSSize(width: bitmap.pixelsWide, height: bitmap.pixelsHigh))
    source.addRepresentation(bitmap)
    NSGraphicsContext.saveGraphicsState()
    guard let context = NSGraphicsContext(bitmapImageRep: scaled) else {
        NSGraphicsContext.restoreGraphicsState()
        return bitmap
    }
    NSGraphicsContext.current = context
    context.imageInterpolation = .high
    source.draw(
        in: NSRect(x: 0, y: 0, width: targetWidth, height: targetHeight),
        from: NSRect(x: 0, y: 0, width: bitmap.pixelsWide, height: bitmap.pixelsHigh),
        operation: .copy,
        fraction: 1
    )
    NSGraphicsContext.restoreGraphicsState()
    return scaled
}

/// Handles screenshot (viewport and full-page).
struct ScreenshotHandler {
    let context: HandlerContext

    func register(on router: Router) {
        router.register("screenshot") { body in await screenshot(body) }
    }

    @MainActor
    private func screenshot(_ body: [String: Any]) async -> [String: Any] {
        let tabId = HandlerContext.tabId(from: body)
        let format = body["format"] as? String ?? "png"
        guard let resolution = ScreenshotResolution.parse(body["resolution"]) else {
            return errorResponse(code: "INVALID_PARAMS", message: "resolution must be 'native' or 'viewport'")
        }

        do {
            _ = try context.resolveRenderer(tabId: tabId)
            let image = try await context.takeSnapshot(tabId: tabId)
            let quality = ((body["quality"] as? NSNumber)?.doubleValue ?? 80) / 100.0
            return successResponse(
                try await context.screenshotPayload(
                    from: image,
                    format: format,
                    quality: quality,
                    resolution: resolution,
                    tabId: tabId
                )
            )
        } catch {
            if let tabError = tabErrorResponse(from: error) { return tabError }
            return errorResponse(code: "SCREENSHOT_FAILED", message: error.localizedDescription)
        }
    }
}

File: apps/android/app/src/main/java/com/kelpie/browser/handlers/ScreenshotHandler.kt
package com.kelpie.browser.handlers

import com.kelpie.browser.network.Router
import com.kelpie.browser.network.errorResponse
import com.kelpie.browser.network.successResponse

enum class ScreenshotResolution(
    val wireValue: String,
) {
    NATIVE("native"),
    VIEWPORT("viewport"),
    ;

    companion object {
        fun parse(raw: Any?): ScreenshotResolution? =
            when (raw) {
                null -> NATIVE
                is String -> entries.firstOrNull { it.wireValue == raw }
                else -> null
            }
    }
}

data class ScreenshotViewportMetrics(
    val viewportWidth: Int,
    val viewportHeight: Int,
    val devicePixelRatio: Double,
) {
    fun metadata(
        imageWidth: Int,
        imageHeight: Int,
        format: String,
        resolution: ScreenshotResolution,
    ): Map<String, Any> {
        val scaleX = if (viewportWidth > 0) imageWidth.toDouble() / viewportWidth.toDouble() else 1.0
        val scaleY = if (viewportHeight > 0) imageHeight.toDouble() / viewportHeight.toDouble() else 1.0
        return mapOf(
            "width" to imageWidth,
            "height" to imageHeight,
            "format" to format,
            "resolution" to resolution.wireValue,
            "coordinateSpace" to "viewport-css-pixels",
            "viewportWidth" to viewportWidth,
            "viewportHeight" to viewportHeight,
            "devicePixelRatio" to devicePixelRatio,
            "imageScaleX" to scaleX,
            "imageScaleY" to scaleY,
        )
    }
}

class ScreenshotHandler(
    private val ctx: HandlerContext,
) {
    fun register(router: Router) {
        router.register("screenshot") { screenshot(it) }
    }

    private suspend fun screenshot(body: Map<String, Any?>): Map<String, Any?> {
        val format = body["format"] as? String ?: "png"
        val quality = (body["quality"] as? Int) ?: 80
        val resolution =
            ScreenshotResolution.parse(body["resolution"])
                ?: return errorResponse("INVALID_PARAMS", "resolution must be 'native' or 'viewport'")
        val payload =
            ctx.captureScreenshotPayload(format, quality, resolution)
                ?: return errorResponse("SCREENSHOT_FAILED", "Failed to capture screenshot")
        return successResponse(payload)
    }
}
